### ISSUE TITLE
Reading  h5ad, where my dimension reduction go?

### ISSUE THREAD

**Issue by Xparx** [satijalab/seurat#1427]
Hi guys,

Seurat version:
[1] Seurat_3.0.0

Trying to convert an object from scanpy to seurat for friends and family to use.

In python the AnnData object looks like
```
AnnData object with n_obs × n_vars = 2562 × 13042 
    obs: 'n_counts', 'n_genes', 'hpf', 'louvain', 'leiden'
    var: 'Gene ID', 'n_cells'
    uns: 'diffmap_evals', 'hpf_colors', 'leiden', 'louvain', 'neighbors', 'pca', 'rank_genes_groups'
    obsm: 'X_pca', 'X_tsne', 'X_umap'
    varm: 'PCs'

adata.write(os.path.join(datadir, "adata_preprocessed_without_raw_pipeline.h5ad"))
```

However when loading it in R i get the following output
```
 bla <- ReadH5AD(file = "adata_preprocessed_without_raw_pipeline.h5ad")
Pulling expression matrices and metadata
Data is unscaled
Creating assay object
Storing X as raw and raw as counts
No variable feature expression found in h5ad file
No dimensional reduction information found
Assembling Seurat object
No nearest-neighbor graph
```
For some reason none of the properties can be found.

It would be great to get some help in fixing this. What should I look for?

**Comment by mojaveazure** [comment]
Hi Andreas,

From the [documentation for `ReadH5AD`](https://www.rdocumentation.org/packages/Seurat/versions/3.0.0/topics/ReadH5AD):

> adata.X is dense and adata.raw is filled [...] Objects will be filled with scaled and normalized data
> ...
> In addition, dimensional reduction information and nearest-neighbor graphs will be searched for and added if and only if scaled data is being added.

What this means is that `ReadH5AD` assumes that if `X` is a sparse matrix, the data isn't scaled and if the data is unscaled, there shouldn't be any dimensional reduction information or nearest neighbor graphs. This is assumed as dimensional reduction and clustering is usually only done on scaled counts and scaled counts are generally stored as a dense matrix, not a sparse one.

If you modify your AnnData object so that `X` is a dense matrix with the scaled counts and store the normalized expression matrix in the `raw.X` slot, everything will fill out as expected. You need both as the Seurat object requires either the `data` slot of each Assay to be filled, and this is usually normalized counts except when first creating the Seurat object.

**Comment by Xparx** [comment]
Great thanks! 
This got me closer I think.
Now It seems to understand and add dimensionality reduction and other features. 
However one more error appears. (Seems to be fixed with an edit in python, see bottom.)

```
> library('Seurat')
> adata = ReadH5AD(file = 'adata_preprocessed_dense_X_pipeline.h5ad')
Pulling expression matrices and metadata
Data is scaled
Merging feature-level metadata dataframes
Creating assay object
Storing X as scale.data and raw as data
Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')
Warning: Adding features not currently present in the object
No variable feature expression found in h5ad file
Pulling dimensional reduction information
Pulling cell embeddings
Pulling feature loadings
Error in dimnames(x) <- dn : 
  length of 'dimnames' [1] not equal to array extent
```

The complete anndata object looks like this
```
AnnData object with n_obs × n_vars = 2562 × 13042 
    obs: 'n_counts', 'n_genes', 'hpf', 'louvain', 'leiden'
    var: 'Gene ID', 'n_cells'
    uns: 'diffmap_evals', 'hpf_colors', 'leiden', 'louvain', 'neighbors', 'pca', 'rank_genes_groups'
    obsm: 'X_pca', 'X_tsne', 'X_umap', 'X_diffmap'
    varm: 'PCs'
```
It has another layer as well but i deleted it as I thought it might be the cause of the error.

Fix?!
------
Actually. It turns out that preprocessing my var_names before with
```
adata.var_names = [i.replace('_', '-') for i in adata.var_names]
```
Seems to have fixed this issue. Even though it complains a bit
```
adata = ReadH5AD(file = 'adata_preprocessed_dense_X_pipeline.h5ad')
Pulling expression matrices and metadata
Data is scaled
Merging feature-level metadata dataframes
Creating assay object
Storing X as scale.data and raw as data
Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')
No variable feature expression found in h5ad file
Pulling dimensional reduction information
Pulling cell embeddings
Pulling feature loadings
Warning: No columnames present in cell embeddings, setting to 'PC_1:50'
Warning: No columnames present in cell embeddings, setting to 'tSNE_1:2'
Warning: No columnames present in cell embeddings, setting to 'UMAP_1:2'
Warning: No columnames present in cell embeddings, setting to 'DIFFMAP_1:15'
Assembling Seurat object
Finding nearest neighbor graph
```

Not sure what fixed what here but I'll add it here if it ever comes up again.

**Comment by mojaveazure** [comment]
Hi Andreas,

That error was happening because we create the Seurat object with either the `counts` or `data` slot first, then add `scale.data`. However, our mechanism for filling the `scale.data` slot would remove features that had an underscore rather than replace the underscore with a dash. That's why it works when you preprocess `var_names` in Python before reading into Seurat.

This behaviour should be fixed as of commit 1f18dd2 on the `develop` branch. To install the development version of Seurat, please see the instructions [here](https://satijalab.org/seurat/install.html#development).

Thanks for bringing this up!

**Comment by screamer** [comment]
Hello guys,

When I use ReadH5AD for a h5ad file, it can't pull the cell emebedding. I don't know the reason. Could anybody help me?

AnnData object with n_obs × n_vars = 22006 × 2000
    obs: 'batch', 'batch_name', 'cell_type', 'cell_type_code', 'n_counts', 'n_genes', 'percent_mito', 'replicate_id'
uns: 'neighbors'
obsm: 'X_scxx', 'X_umap'

seurat 3.0.2

ReadH5AD(file = filename)
Pulling expression matrices and metadata
Data is scaled
Merging feature-level metadata dataframes
Creating assay object
Storing X as scale.data and raw as data
Warning: Adding features not currently present in the object
No variable feature expression found in h5ad file
Pulling dimensional reduction information
Pulling cell embeddings
Error in file[["obsm"]]$get_type() : attempt to apply non-function
